amaranth/hdl/_ast.py

```diff
@@ -2732,6 +2742,160 @@ def case_repr(keys, stmts):
         return "(switch {!r} {})".format(self.test, " ".join(case_reprs))
 
 
+class IOValue(metaclass=ABCMeta):
+    @staticmethod
+    def cast(obj):
+        if isinstance(obj, IOValue):
+            return obj
+        elif isinstance(obj, Value) and len(obj) == 0:
+            return IOCat()
+        else:
+            raise TypeError(f"Object {obj!r} cannot be converted to an IO value")
+
+    def __init__(self, *, src_loc_at=0):
+        super().__init__()
+        self.src_loc = tracer.get_src_loc(1 + src_loc_at)
+
+    @property
+    @abstractmethod
+    def metadata(self):
+        raise NotImplementedError # :nocov:
+
+    def __getitem__(self, key):
+        n = len(self)
+        if isinstance(key, int):
+            if key not in range(-n, n):
+                raise IndexError(f"Index {key} is out of bounds for a {n}-bit IO value")
+            if key < 0:
+                key += n
+            return IOSlice(self, key, key + 1, src_loc_at=1)
+        elif isinstance(key, slice):
+            start, stop, step = key.indices(n)
+            if step != 1:
+                return IOCat((self[i] for i in range(start, stop, step)), src_loc_at=1)
+            return IOSlice(self, start, stop, src_loc_at=1)
+        else:
+            raise TypeError(f"Cannot index IO value with {key!r}")
+
+    @abstractmethod
+    def _ioports(self):
+        raise NotImplementedError # :nocov:
+
+
+@final
+class IOPort(IOValue):
+    def __init__(self, width, *, name=None, attrs=None, metadata=None, src_loc_at=0):
+        super().__init__(src_loc_at=src_loc_at)
+
+        if name is not None and not isinstance(name, str):
+            raise TypeError(f"Name must be a string, not {name!r}")
+        self.name = name or tracer.get_var_name(depth=2 + src_loc_at)
+
+        self._width = operator.index(width)
+        self._attrs = dict(() if attrs is None else attrs)
+        self._metadata = (None,) * self._width if metadata is None else tuple(metadata)
+
+    def __len__(self):
+        return self._width
+
+    @property
+    def width(self):
+        return self._width
+
+    @property
+    def attrs(self):
+        return self._attrs
+
+    @property
+    def metadata(self):
+        return self._metadata
+
+    def _ioports(self):
+        return {self}
+
+    def __repr__(self):
+        return f"(ioport {self.name})"
```

Comment:
Maybe with a dash, to visually group the `io-` variants?

```suggestion
        return f"(io-port {self.name})"
```